vortex-serde/src/layouts/read/mask.rs

```diff
@@ -158,12 +158,20 @@ impl RowMask {
             return Ok(Some(sliced.clone()));
         }
 
-        let predicate = self.to_predicate_array()?;
+        if (true_count as f64 / sliced.len() as f64) < (1.0 / 1024.0) {
+            let indices = self.to_indices_array()?;
+            take(sliced, indices).map(Some)
+        } else {
+            let mask = self.to_mask_array()?;
+            filter(sliced, mask).map(Some)
+        }
+    }
 
-        filter(sliced, predicate).map(Some)
+    pub fn to_indices_array(&self) -> VortexResult<Array> {
+        Ok(PrimitiveArray::from_vec(self.values.to_vec(), Validity::NonNullable).into_array())
```

Comment:
I guess we actually need roaring arrays for this to avoid the copy here

Reply:
Oh let me try using roaring int

Reply:
yes sorry you do need a bit more specialization across the board. Also roaring int array has serialization cost. This is not a simple fix but potentially perf